src/SegmentedArray.chpl

```diff
@@ -75,16 +70,228 @@ module SegmentedArray {
+                    for pos in 0..#gl {
+                        gatheredVals[go+pos] = va[oa[idx:int]+pos];
+                    }
+                }
+            }
+            saLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),
+                                    "Gathered offsets and vals in %i seconds".format(
+                                                getCurrentTime() -t1));
+            return (gatheredOffsets, gatheredVals);
+        }
+
+        /* Logical indexing (compress) of SegArray. */
+        proc this(iv: [?D] bool) throws {
+            // Index vector must be same domain as array
+            if (D != segments.aD) {
+                saLogger.info(getModuleName(),getRoutineName(),getLineNumber(),
+                                                                "Array out of bounds");
+                throw new owned OutOfBoundsError();
+            }
+            saLogger.debug(getModuleName(),getRoutineName(),getLineNumber(), 
+                                                        "Computing lengths and offsets");
+
+            ref oa = segments.a;
+            const low = segments.aD.low, high = segments.aD.high;
+            // check there's enough room to create a copy for scan and throw if creating a copy would go over memory limit
+            overMemLimit(numBytes(int) * iv.size);
+            // Calculate the destination indices
+            var steps = + scan iv;
+            var newSize = steps[high];
+            steps -= iv;
+            // Early return for zero-length result
+            if (newSize == 0) {
+                return (makeDistArray(0, int), makeDistArray(0, values.etype));
+            }
+            var segInds = makeDistArray(newSize, int);
+            forall (t, dst, idx) in zip(iv, steps, D) with (var agg = newDstAggregator(int)) {
+                if t {
+                    agg.copy(segInds[dst], idx);
+                }
+            }
+            return this[segInds];
         }
 
         proc getNonEmpty() throws {
-            return getLengths() > 0;
+            return lengths > 0;
         }
 
         proc getNonEmptyCount() throws {
             var non_empty = getNonEmpty();
             return + reduce non_empty:int;
         }
+
+        proc getValuesName(st: borrowed SymTab): string throws {
+            var vname = st.nextName();
+            st.addEntry(vname, values);
+            return vname;
+        }
+
+        proc getSegmentsName(st: borrowed SymTab): string throws {
+            var sname = st.nextName();
+            st.addEntry(sname, segments);
+            return sname;
+        }
+
+        proc getLengthsName(st: borrowed SymTab): string throws {
+            var lname = st.nextName();
+            st.addEntry(lname, new shared SymEntry(lengths));
+            return lname;
+        }
+
+        proc buildReturnMap(ref rm: map(string, string), st: borrowed SymTab) throws {
```

Comment:
this is kinda nitpicky but a name like `populate/fillReturnMap` would be more descriptive

Reply:
done